File: waltid-applications/waltid-wallet-demo-compose/sharedLogic/src/commonMain/kotlin/id/walt/walletdemo/compose/logic/VerifierClientIdentifier.kt
package id.walt.walletdemo.compose.logic

internal data class VerifierClientIdentifier(
    val scheme: Scheme,
    val value: String,
    val rawValue: String,
) {
    enum class Scheme(
        val prefix: String?,
        val genericLabel: String?,
    ) {
        RedirectUri(prefix = "redirect_uri", genericLabel = null),
        X509SanDns(prefix = "x509_san_dns", genericLabel = null),
        X509Hash(prefix = "x509_hash", genericLabel = "X.509 verifier"),
        DecentralizedIdentifier(prefix = "decentralized_identifier", genericLabel = "DID verifier"),
        LegacyDid(prefix = "did", genericLabel = "DID verifier"),
        VerifierAttestation(prefix = "verifier_attestation", genericLabel = "Verifier attestation"),
        OpenIdFederation(prefix = "openid_federation", genericLabel = "OpenID Federation verifier"),
        PreRegistered(prefix = null, genericLabel = null),
        Unsupported(prefix = null, genericLabel = null),
    }

    companion object {
        fun parse(rawClientId: String): VerifierClientIdentifier? {
            val rawValue = rawClientId.trim().takeIf { it.isNotBlank() } ?: return null
            if (NetworkOrigin.parse(rawValue) != null) {
                return VerifierClientIdentifier(scheme = Scheme.PreRegistered, value = rawValue, rawValue = rawValue)
            }

            val prefixedValue = ClientIdentifierPrefix.split(rawValue)
            if (prefixedValue == null) {
                return VerifierClientIdentifier(scheme = Scheme.PreRegistered, value = rawValue, rawValue = rawValue)
            }

            val scheme = Scheme.entries.firstOrNull { it.prefix == prefixedValue.prefix }
                ?: Scheme.Unsupported

            return VerifierClientIdentifier(scheme = scheme, value = prefixedValue.value, rawValue = rawValue)
        }
    }
}

private data class ClientIdentifierPrefix(
    val prefix: String,
    val value: String,
) {
    companion object {
        fun split(rawValue: String): ClientIdentifierPrefix? {
            val separatorIndex = rawValue.indexOf(':')
            if (separatorIndex <= 0) return null

            return ClientIdentifierPrefix(
                prefix = rawValue.take(separatorIndex),
                value = rawValue.drop(separatorIndex + 1),
            )
        }
    }
}

File: waltid-applications/waltid-wallet-demo-ios/iosApp/iosApp/Display/VerifierClientIdentifier.swift
import Foundation

struct VerifierClientIdentifier {
    let scheme: Scheme
    let value: String
    let rawValue: String

    enum Scheme: CaseIterable {
        case redirectURI
        case x509SanDNS
        case x509Hash
        case decentralizedIdentifier
        case legacyDID
        case verifierAttestation
        case openIDFederation
        case preRegistered
        case unsupported

        var prefix: String? {
            switch self {
            case .redirectURI: return "redirect_uri"
            case .x509SanDNS: return "x509_san_dns"
            case .x509Hash: return "x509_hash"
            case .decentralizedIdentifier: return "decentralized_identifier"
            case .legacyDID: return "did"
            case .verifierAttestation: return "verifier_attestation"
            case .openIDFederation: return "openid_federation"
            case .preRegistered, .unsupported: return nil
            }
        }

        var genericLabel: String? {
            switch self {
            case .x509Hash: return "X.509 verifier"
            case .decentralizedIdentifier, .legacyDID: return "DID verifier"
            case .verifierAttestation: return "Verifier attestation"
            case .openIDFederation: return "OpenID Federation verifier"
            case .redirectURI, .x509SanDNS, .preRegistered, .unsupported: return nil
            }
        }
    }

    static func parse(_ rawClientID: String) -> VerifierClientIdentifier? {
        let rawValue = rawClientID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !rawValue.isEmpty else { return nil }

        if NetworkOrigin.parse(rawValue) != nil {
            return VerifierClientIdentifier(scheme: .preRegistered, value: rawValue, rawValue: rawValue)
        }

        guard let prefixedValue = ClientIdentifierPrefix.split(rawValue) else {
            return VerifierClientIdentifier(scheme: .preRegistered, value: rawValue, rawValue: rawValue)
        }

        let scheme = Scheme.allCases.first { $0.prefix == prefixedValue.prefix } ?? .unsupported

        return VerifierClientIdentifier(scheme: scheme, value: prefixedValue.value, rawValue: rawValue)
    }
}

private struct ClientIdentifierPrefix {
    let prefix: String
    let value: String

    static func split(_ rawValue: String) -> ClientIdentifierPrefix? {
        guard let separator = rawValue.firstIndex(of: ":"),
              separator != rawValue.startIndex else {
            return nil
        }

        return ClientIdentifierPrefix(
            prefix: String(rawValue[..<separator]),
            value: String(rawValue[rawValue.index(after: separator)...])
        )
    }
}

struct NetworkOrigin {
    let host: String

    static func parse(_ rawValue: String) -> NetworkOrigin? {
        let value = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty,
              let url = URL(string: value),
              let scheme = url.scheme?.lowercased(),
              httpSchemes.contains(scheme),
              let host = url.host,
              !host.isEmpty else {
            return nil
        }

        return NetworkOrigin(host: host)
    }

    private static let httpSchemes: Set<String> = ["http", "https"]
}
